packages/file-search/src/node/file-search-service-impl.ts

```diff
@@ -0,0 +1,106 @@
+
+@injectable()
+export class FileSearchServiceImpl implements FileSearchService {
+
+    constructor( @inject(ILogger) private logger: ILogger) { }
+
+    async find(uri: string, searchPattern: string, options?: { fuzzyMatch?: boolean, limit?: number, useGitignore?: boolean }): Promise<string[]> {
+        const basePath = FileUri.fsPath(new URI(uri));
+        const opts = {
+            fuzzyMatch: true,
+            limit: Number.MAX_SAFE_INTEGER,
+            useGitignore: true,
+            ...options
+        };
+        const result: string[] = [];
+        const limitReached = new Error("limit reached");
+        const filtered = this.getFiltered(basePath, opts);
+        try {
+            this.findRecursive(basePath, filtered, filePath => {
+                if (result.length >= opts.limit) {
+                    throw limitReached;
+                }
+                if (opts.fuzzyMatch && fuzzy.test(searchPattern, filePath)) {
+                    result.push(FileUri.create(filePath).toString());
+                } else {
+                    if (filePath.toLocaleLowerCase().indexOf(searchPattern.toLocaleLowerCase()) !== -1) {
+                        result.push(FileUri.create(filePath).toString());
+                    }
+                }
+            });
+        } catch (e) {
+            if (e !== limitReached) {
+                throw e;
+            }
+        }
+        return result;
+    }
+
+    private findRecursive(filePath: string, filtered: Filtered, acceptor: (fileName: string) => void) {
+        const result = fs.readdirSync(filePath);
+        for (const child of result) {
+            const childPath = path.join(filePath, child);
+            if (!filtered(child, childPath)) {
+                if (fs.statSync(childPath).isDirectory()) {
+                    this.findRecursive(childPath, filtered, acceptor);
+                } else {
+                    acceptor(childPath);
+                }
+            }
+        }
+    }
+
+    private getFiltered(basePath: string, options: { fuzzyMatch: boolean, limit: number, useGitignore: boolean }): Filtered {
+        const defaultFilter = (simpleName: string) =>
+            simpleName === '.git' ||
+            simpleName.startsWith('.git/');
+        if (options.useGitignore) {
+            const gitIgnore = this.findGitIgnore(basePath);
+            if (gitIgnore) {
+                const matcher = parser.compile(gitIgnore.contents);
+                return (short, long) => defaultFilter(short) || matcher.denies(short);
+            }
+        }
+        return defaultFilter;
+    }
+
+    private findGitIgnore(basePath: string): { path: string, contents: string } | undefined {
+        try {
+            const fullPath = path.join(basePath, '.gitignore');
+            const result = fs.readFileSync(fullPath, 'utf-8');
+            if (this.logger.isInfo()) {
+                this.logger.info(`Found gitignore below ${fullPath}`);
```

Comment:
Sounds more like a debug log and you don't have to check for isInfo, .info will do it for you.

Reply:
It avoids string concatenation. This is not a hot-spot, though, but I use that pattern generally.